### ISSUE TITLE
[QUERY] Writing unit tests for class with `BlobContainerClient`. It's not mockable?

### ISSUE THREAD

**Issue by dinopraso** [Azure/azure-sdk-for-java#21849]
**Query/Question**
Currently, the main way of interacting with the Azure Blob Storage service is through the `BlobContainerClient` class.
In my project, we have our own data storage abstractions in place, and one of the implementations of those uses Azure Blob Storage. The issue presented itself when I started to write the first unit test for that implementation when I realised that the `BlobContainerClient` class is `final` and thus cannot be mocked by either `Mockito` nor `EasyMock`, which is a big issue, especially since `BlobContainerClient` methods are not exposed via any interface.

How am I to write unit tests for this? One idea might be to wrap the Azure class in another wrapper which could be mocked, but that is awfully inconvenient and really just passes the issue down to "how to test the new wrapper?".

I really hope that there is some other solution here which I am overlooking, and hope that this wasn't an oversight.

***Why is this not a Bug or a feature Request?***
I hope it's not an oversight and that there is another solution here.

**Setup (please complete the following information if applicable):**
 - OS: [e.g. iOS]: N/A
 - IDE : [e.g. IntelliJ]: N/A
 - Version of the Library used: 12.10.2
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Query Added
- [x] Setup information Added

**Comment by joshfree** [comment]
@gapra-msft could you please follow up with @dinopraso

**Comment by gapra-msft** [comment]
Hi @dinopraso, Thank you for your question. 

By design, we make most of our client classes final. [Here](https://github.com/Azure/azure-sdk-for-java/wiki/Unit-Testing#mocking-final-classes) is some more information on how to Mock final classes. Please let us know if you hit any further issues.

**Comment by dinopraso** [comment]
@gapra-msft Thank you for link. I failed to find that documentation page on my own. I was hoping not to have to resort to mocking extensions which may well become technical debt in the near future considering the increased focus on locking down reflection access to foreign modules in Java. 

Additionally, while I fully appreciate the fact that everyone should abstract 3rd party libraries (like we are indeed doing) still doesn't really change the fact that the concrete implementation which uses the BlobContainerClient will remain either untested or incur immediate technical debt because of its reliance on circumventing final class modifiers.

Nevertheless, I appreciate your response. I would, out of curiosity, like to know why such a design decision was made instead of exploding an interface like most other 3rd party libraries and SDKs?

**Comment by gapra-msft** [comment]
Tagging @JonathanGiles, @srnagar, @alzimmermsft for further insight on why the design decision to make client classes final was made.

**Comment by srnagar** [comment]
@dinopraso Thank you for your feedback. The reason for making the client classes final is to not eagerly expose the service clients for extension unless it was required. This is to reduce the possibility of a breaking change when newer versions of the client are released with behavioral changes that can break subclasses. Using interface for clients would require a `default` implementation for each new method added to the client after the first release.

Also, making a `final` class non-final is not a breaking change. So, we started with making the clients `final` with the possibility of removing `final` if we got customer feedback to do so. 

We are discussing this with the team to decide if we should remove `final` for client classes. I will update this issue with what we decide to do going forward.
